File: chain-registry-types/ts/example/fetch-chain-data.js
import { fetchChain, fetchAssetList, fetchVersions, fetchIBCData } from "@puneet_m/chain-registry-interface";

async function fetchChainData() {
  try {
    console.log("=== Fetching Osmosis Chain Data ===");
    
    // Fetch chain information
    console.log("Fetching chain information...");
    const chainInfo = await fetchChain("osmosis");
    console.log(`Chain Name: ${chainInfo.chain_name}`);
    console.log(`Chain ID: ${chainInfo.chain_id}`);
    console.log(`Bech32 Prefix: ${chainInfo.bech32_prefix}`);
    console.log(`Status: ${chainInfo.status}`);
    console.log(`Network Type: ${chainInfo.network_type || 'mainnet'}`);
    console.log(`Pretty Name: ${chainInfo.pretty_name}`);
    console.log(`Website: ${chainInfo.website}`);
    console.log(`Description: ${chainInfo.description}`);
    
    // Fetch asset list
    console.log("\nFetching asset list...");
    const assetList = await fetchAssetList("osmosis");
    console.log(`Number of assets: ${assetList.assets.length}`);
    
    // Show first few assets
    console.log("\nFirst 3 assets:");
    for (let i = 0; i < Math.min(3, assetList.assets.length); i++) {
      const asset = assetList.assets[i];
      console.log(`  - Name: ${asset.name}, Symbol: ${asset.symbol}, Base: ${asset.base}`);
    }
    
    // Show native currency details
    console.log("\nNative Currency Information:");
    console.log(`  Base Denom: ${assetList.assets[0].base}`);
    console.log(`  Display Denom: ${assetList.assets[0].display}`);
    console.log(`  Symbol: ${assetList.assets[0].symbol}`);
    
    // Show chain features
    console.log("\nChain Features:");
    console.log(`  Daemon Name: ${chainInfo.daemon_name}`);
    console.log(`  Node Home: ${chainInfo.node_home}`);
    console.log(`  Key Algorithms: ${chainInfo.key_algos?.join(', ')}`);
    console.log(`  SLIP44: ${chainInfo.slip44}`);
    
    // Show API endpoints
    console.log("\nAPI Endpoints:");
    if (chainInfo.apis?.rpc && chainInfo.apis.rpc.length > 0) {
      console.log(`  RPC: ${chainInfo.apis.rpc[0].address}`);
    }
    if (chainInfo.apis?.rest && chainInfo.apis.rest.length > 0) {
      console.log(`  REST: ${chainInfo.apis.rest[0].address}`);
    }
    if (chainInfo.apis?.grpc && chainInfo.apis.grpc.length > 0) {
      console.log(`  gRPC: ${chainInfo.apis.grpc[0].address}`);
    }
    
    // Fetch versions information
    console.log("\nFetching versions information...");
    try {
      const versions = await fetchVersions("osmosis");
      console.log("Versions information:");
      console.log(`  Chain Name: ${versions.chain_name}`);
      console.log(`  Number of versions: ${versions.versions.length}`);
      if (versions.versions.length > 0) {
        const latestVersion = versions.versions[0];
        console.log(`  Latest version: ${latestVersion.name}`);
        if (latestVersion.tag) console.log(`  Tag: ${latestVersion.tag}`);
        if (latestVersion.height) console.log(`  Height: ${latestVersion.height}`);
      }
    } catch (versionError) {
      console.log(`  Could not fetch versions: ${versionError.message}`);
    }

    // Fetch IBC data information
    console.log("\nFetching IBC data information...");
    try {
      const ibcData = await fetchIBCData();
      console.log("IBC Data information:");
      console.log(`  Schema: ${ibcData["$schema"]}`);
      console.log(`  Total IBC records: ${Object.keys(ibcData).filter(key => key !== '$schema').length}`);
      console.log("  Note: The published package has fetchIBCData but not fetchIBCConnection for specific chain pairs");
    } catch (ibcError) {
      console.log(`  Could not fetch IBC data: ${ibcError.message}`);
      console.log("  Note: IBC data might not be available or accessible");
    }

    console.log("\n=== Chain Data Fetch Complete ===");
  } catch (error) {
    console.error("Error fetching chain data:", error);
  }
}

// Run the example
fetchChainData().catch(console.error);

File: chain-registry-types/rust/example/fetch-chain-data.rs
use chain_registry_interface::{fetch_chain, fetch_asset_list, fetch_versions, fetch_ibc_connection};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    fetch_chain_data().await?;
    Ok(())
}

async fn fetch_chain_data() -> Result<(), Box<dyn std::error::Error>> {
    println!("=== Fetching Osmosis Chain Data ===");

    // Fetch chain information
    println!("Fetching chain information...");
    let chain = fetch_chain("osmosis").await?;
    println!("Chain Name: {}", chain.chain_name);
    println!("Chain Type: {}", chain.chain_type);
    println!("Status: {}", chain.status);

    // Fetch asset list
    println!("\nFetching asset list...");
    let assets = fetch_asset_list("osmosis").await?;
    println!("Number of assets: {}", assets.assets.len());

    // Show first few assets
    println!("\nFirst 3 assets:");
    for i in 0..std::cmp::min(3, assets.assets.len()) {
        let asset = &assets.assets[i];
        println!("  - Name: {}, Symbol: {}, Base: {}",
                 asset.name,
                 asset.symbol,
                 asset.base);
    }

    // Show native currency details
    println!("\nNative Currency Information:");
    if !assets.assets.is_empty() {
        let first_asset = &assets.assets[0];
        println!("  Base Denom: {}", first_asset.base);
        println!("  Display Denom: {}", first_asset.display);
        println!("  Symbol: {}", first_asset.symbol);
    }

    // Show chain features
    println!("\nChain Features:");
    println!("  Chain Type: {}", chain.chain_type);
    println!("  Status: {}", chain.status);

    // Fetch versions information
    println!("\nFetching versions information...");
    match fetch_versions("osmosis").await {
        Ok(versions) => {
            println!("Versions information:");
            println!("  Chain Name: {}", versions.chain_name);
            println!("  Number of versions: {}", versions.versions.len());
            if !versions.versions.is_empty() {
                let latest_version = &versions.versions[0];
                println!("  Latest version: {}", latest_version.name);
                if let Some(ref tag) = latest_version.tag {
                    println!("  Tag: {}", tag);
                }
                if let Some(height) = latest_version.height {
                    println!("  Height: {}", height);
                }
            }
        }
        Err(e) => {
            println!("  Could not fetch versions: {}", e);
        }
    }

    // Fetch IBC connection information (between osmosis and cosmoshub)
    println!("\nFetching IBC connection information...");
    match fetch_ibc_connection("osmosis", "cosmoshub").await {
        Ok(ibc_connection) => {
            println!("IBC Connection information:");
            println!("  Chain 1: {} ({})", ibc_connection.chain_1.chain_name, ibc_connection.chain_1.chain_id);
            println!("  Chain 2: {} ({})", ibc_connection.chain_2.chain_name, ibc_connection.chain_2.chain_id);
            println!("  Number of channels: {}", ibc_connection.channels.len());
            if !ibc_connection.channels.is_empty() {
                let first_channel = &ibc_connection.channels[0];
                println!("  First channel: {} <-> {}", first_channel.chain_1.channel_id, first_channel.chain_2.channel_id);
                println!("  Ordering: {}", first_channel.ordering);
                println!("  Version: {}", first_channel.version);
            }
        }
        Err(e) => {
            println!("  Could not fetch IBC connection: {}", e);
            println!("  Note: Not all chain pairs have IBC connections defined in the registry");
        }
    }

    println!("\n=== Chain Data Fetch Complete ===");

    Ok(())
}